### DISCUSSION TITLE
how to write block handler ?

### DISCUSSION THREAD

**Discussion by begemot63** [alibaba/Sentinel#2148] [Q&A]
1. When blockHandler() method will be called?
2. I want to write block handler . How I can do it?
The handler needs to return null object to method caller if  exception was caught.

**Comment by sczyh30** [comment]
Hi, if you're referring block handler and fallback in `@SentinelResource` annotation, you may refer to https://github.com/alibaba/Sentinel/tree/master/sentinel-extension/sentinel-annotation-aspectj

**Comment by begemot63** [comment]
Hi Eric
I read that document.

However my code annotation did not work.
I  used this code:

public class MemcachedCache extends AbstractValueAdaptingCache {

    @Override
    @SentinelResource(value = MEMCACHED_RESILIENCE_RESOURCE,
fallback = "fallbackForLookup")
    public Object lookup(Object key) {
      ...
      // memcached calls
    }

Note, the class which shown here MemcachedCache extends
AbstractValueAdaptingCache

which implements Cache interface.

The class   MemcachedCache used  as implementation of Cacha interface
to support Cachable

spring boot annotation. So the loookup(), get(), put() method invoked
by spring boot framework.

It looks like your proxy pattern with  @SentinelResource does not work.

So I decided to use the  Sentinel  SphU  API, (Not sure what SphU stands for , but) I have the
following code now:

public Object lookup(Object key) {
  ...
  Entry entry = null;
  try {
    entry = SphU.entry(MEMCACHED_RESILIENCE_RESOURCE, EntryType.IN, 1, key);

      .....
     // access to memcached
     // to simulate error:
      throw new RuntimeException("error");

  } catch (BlockException ex) {
            LOGGER.info("lookup, Cuitcut Breaker is open", name);
            handleMetrics("lookup",  name, key.toString(), memCachedKey, ex);
            return null;
  } catch (Throwable th) {
            Tracer.trace(th);
            LOGGER.info("lookup, Remote cache '{}' is not available", name);
            ClusterNode createRoleNode = ClusterBuilderSlot.getClusterNode("memcached");
            LOGGER.info("lookup, Exception Total Count: {}", createRoleNode.totalException());
            LOGGER.info("lookup, Block Request Count: {}", createRoleNode.blockRequest());
            LOGGER.info("lookup, DegradeRuleManager.getRules(): {} ",DegradeRuleManager.getRules());
            handleMetrics("lookup",  name, key.toString(), memCachedKey, th);
            return null;
   } finally {
            if(entry!=null) {
                entry.exit();
            }
     }

 
Log indicates I have defined 2  DegradeRule
   DegradeRule{resource=memcached, grade=1, count=0.6, limitApp=default, timeWindow=20, minRequestAmount=5, slowRatioThreshold=1.0, statIntervalMs=1000}, 
   DegradeRule{resource=memcached, grade=2, count=10.0, limitApp=default, timeWindow=20, minRequestAmount=5, slowRatioThreshold=1.0, statIntervalMs=10}]  
Also log shows  Exception Total Count: 11 however  SphU.entry never throws BlockException exception. It means socket breaker was never opened . 

Did I missed something ?

Regards
Vlad

**Reply by SomebodyHuang2017** [reply]
There are several steps in the troubleshooting method：
1. Make sure that the Spring AOP aspect can cover this class
2. If the lookup method is called in abstract class, then you should ensure that it calls the lookup method of the proxy class to take effect. 

such as:
`@Configuration
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class AopConfiguration {

    @Bean
    public OceanusAspect oceanusAspect() {
        return new OceanusAspect();
    }
}`

**Comment by begemot63** [comment]
Hi Eric
The second approach also does not work 
The rules are not charged
The exception counter is increased but BlockException was never thrown
 How to properly apply DegradeRule ?

**Comment by sczyh30** [comment]
> It looks like your proxy pattern with  `@SentinelResource` does not work.

Please check whether your class is a Spring bean (Spring AOP proxy requires the instance managed by Spring).

> The exception counter is increased but BlockException was never thrown

Take rule1 as an example: `DegradeRule{resource=memcached, grade=1, count=0.6, limitApp=default, timeWindow=20, minRequestAmount=5, slowRatioThreshold=1.0, statIntervalMs=1000}`

This indicates: If your QPS > 5, error ratio > 60% in this second, the circuit breaker will be open. Please check whether your requests satisfy the condition.

**Comment by begemot63** [comment]
Hi Eric, 
As you may see  my code  throw news RuntimeException("error");   every time we entry in  the try block.
I remove  DegradeRule{resource=memcached, grade=1, count=0.6, limitApp=default, timeWindow=20, minRequestAmount=5, slowRatioThreshold=1.0, statIntervalMs=1000}
and kept only  second rule:   DegradeRule{resource=memcached, grade=2, count=4.0, limitApp=default, timeWindow=10, minRequestAmount=5, slowRatioThreshold=1.0, statIntervalMs=1000}] 
The method is called hundred times in the loop but BlockException is never thrown.
If I add FlowRule :  FlowRule{resource=memcached, limitApp=default, grade=1, count=1.0, strategy=0, refResource=null, controlBehavior=0, warmUpPeriodSec=10, maxQueueingTimeMs=500, clusterMode=false, clusterConfig=null, controller=com.alibaba.csp.sentinel.slots.block.flow.controller.DefaultController@4a110bc1}
 then the BlockExceptopn is thrown each time when method is called. The flow rule works only if count set to 1 , bigger value does not throw BlockException

**Comment by begemot63** [comment]
I solved the problem by increasing statIntervalMs
